conjure-http/src/server/runtime.rs

```diff
@@ -0,0 +1,391 @@
+
+use crate::server::Encoding;
+use crate::server::JsonEncoding;
+use crate::server::SmileEncoding;
+use conjure_error::Error;
+use conjure_error::InvalidArgument;
+use http::header::ACCEPT;
+use http::header::CONTENT_TYPE;
+use http::HeaderMap;
+use mediatype::names;
+use mediatype::MediaType;
+use mediatype::MediaTypeList;
+use mediatype::ReadParams;
+
+/// A type providing server logic that is configured at runtime.
+pub struct ConjureRuntime {
+    encodings: Vec<Box<dyn Encoding + Sync + Send>>,
+}
+
+impl ConjureRuntime {
+    /// Creates a new runtime with default settings.
+    pub fn new() -> Self {
+        Self::builder().build()
+    }
+
+    /// Creates a new builder.
+    pub fn builder() -> Builder {
+        Builder { encodings: vec![] }
+    }
+
+    /// Returns the appropriate [`Encoding`] to deserialize the request body.
+    ///
+    /// The implementation currently compares the request's `Content-Type` header against [`Encoding::content_type`],
+    /// ignoring parameters.
+    pub fn request_body_encoding(
+        &self,
+        headers: &HeaderMap,
+    ) -> Result<&(dyn Encoding + Sync + Send), Error> {
+        let Some(content_type) = headers.get(CONTENT_TYPE) else {
+            return Err(Error::service_safe(
+                "Content-Type header missing from request",
+                InvalidArgument::new(),
+            ));
+        };
+
+        let content_type = content_type
+            .to_str()
+            .map_err(|e| Error::service_safe(e, InvalidArgument::new()))?;
+        let content_type = MediaType::parse(content_type)
+            .map_err(|e| Error::service_safe(e, InvalidArgument::new()))?;
+
+        self.encodings
+            .iter()
+            .map(|e| &**e)
+            .find(|e| mime_matches(&content_type, *e))
+            .ok_or_else(|| {
+                Error::service_safe(
+                    "request Content-Type not accepted by any encoding",
+                    InvalidArgument::new(),
+                )
+            })
+    }
+
+    /// Returns the appropriate [`Encoding`] to serialize the response body.
+    ///
+    /// The MIME types in the request's `Accept` header are processed in accordance with [RFC 9110]. If two MIME types
+    /// have equal preference by that algorithm, the implementation will prefer the type declared earlier in the header.
+    ///
+    /// MIME types are matched against [`Encoding::content_type`], ignoring parameters. If multiple [`Encoding`]s are
+    /// matched against the same MIME type, the encoding registered first by [`Builder::encoding`] will be selected.
+    ///
+    /// The implementation treats the absense of the `Accept` header equivalently to `Accept: */*`.
+    ///
+    /// [RFC 9110]: https://httpwg.org/specs/rfc9110.html#field.accept
+    pub fn response_body_encoding(
+        &self,
+        headers: &HeaderMap,
+    ) -> Result<&(dyn Encoding + Sync + Send), Error> {
+        let mut types = headers
```

Comment:
conjure-java has much simpler logic for this but I figured we may as well support the RFC-mandated behavior more closely. We still ignore parameters though.